pkgdb/src/buildenv/realise.cc

```diff
@@ -60,59 +56,397 @@ namespace flox::buildenv {
+# variable, and if it hasn't then we'll prepend it to the list and set
+# all the other related env variables.
+declare -a flox_env_dirs
+IFS=: read -ra flox_env_dirs <<< "${FLOX_ENV_DIRS_activate}"
+declare -i flox_env_found=0
+for d in "${flox_env_dirs[@]}"; do
+  if [ "$d" = "$FLOX_ENV" ]; then
+    flox_env_found=1
+    break
+  fi
+done
+if [ $flox_env_found -eq 0 ]; then
+
+  # First activation of this environment. Snapshot environment to start.
+  _start_env="$($_coreutils/bin/mktemp --suffix=.start-env)"
+  export | $_coreutils/bin/sort > "$_start_env"
+
+  # Capture PID of this "first" activation. This provides the unique
+  # identifier with which to refer to environment variables associated
+  # with this environment activation.
+  FLOX_ENV_PID="$$"
+
+  # Set environment variables which represent the cumulative layering
+  # of flox environments. For the most part this involves prepending
+  # to the existing variables of the same name.
+  # TODO: reconcile with CLI which should be setting these. Setting
+  #       "*_activate" variables to indicate the ones we've seen and
+  #       processed on the activate script side, and ultimately also
+  #       for testing/comparison against the CLI-maintained equivalents.
+  FLOX_ENV_DIRS_activate="$FLOX_ENV${FLOX_ENV_DIRS_activate:+:$FLOX_ENV_DIRS_activate}"
+  FLOX_ENV_LIB_DIRS_activate="$FLOX_ENV/lib${FLOX_ENV_LIB_DIRS_activate:+:$FLOX_ENV_LIB_DIRS_activate}"
+  FLOX_PROMPT_ENVIRONMENTS_activate="$FLOX_ENV_DESCRIPTION${FLOX_PROMPT_ENVIRONMENTS_activate:+ $FLOX_PROMPT_ENVIRONMENTS_activate}"
+  export FLOX_ENV_DIRS_activate FLOX_ENV_LIB_DIRS_activate FLOX_PROMPT_ENVIRONMENTS_activate
+
+  # Process the flox environment customizations, which includes (amongst
+  # other things) prepending this environment's bin directory to the PATH.
+  if [ -d "$FLOX_ENV/etc/profile.d" ]; then
+    declare -a _prof_scripts;
+    _prof_scripts=( $(
+      cd "$FLOX_ENV/etc/profile.d";
+      shopt -s nullglob;
+      echo *.sh;
+    ) );
+    for p in "${_prof_scripts[@]}"; do . "$FLOX_ENV/etc/profile.d/$p"; done
+    unset _prof_scripts;
+  fi
+
+  # Set static environment variables from the manifest.
+  if [ -f "$FLOX_ENV/activate.d/envrc" ]; then
+    source "$FLOX_ENV/activate.d/envrc"
+  fi
+
+  # Source the hook-on-activate script if it exists.
+  if [ -e "$FLOX_ENV/activate.d/hook-on-activate" ]; then
+    # Nothing good can come from output printed to stdout in the
+    # user-provided hook scripts because these can get interpreted
+    # as configuration statements by the "in-place" activation
+    # mode. So, we'll redirect stdout to stderr.
+    source "$FLOX_ENV/activate.d/hook-on-activate" 1>&2
+  fi
+
+  # Capture ending environment.
+  _end_env="$($_coreutils/bin/mktemp --suffix=.$FLOX_ENV_PID.end-env)"
+  export | $_coreutils/bin/sort > "$_end_env"
+
+  # The userShell initialization scripts that follow have the potential to undo
```

Comment:
It's important to recall that the user's environment modifications will have already been applied by the time that flox runs, and will therefore be captured in any environment "replay" actions that flox performs.

Consider the various modes:
- interactive: in this case someone is activating an environment from a shell that will have already been customized by user RC files, so we can be confident that we're not tampering with any variables that aren't specifically required by flox
- in-place: in this case flox will be activated from within the user's own RC file(s), and they therefore bear responsibility for placing that activation after whatever customizations they wish to perform have been completed, so therefore it's again the case that flox can assume that the necessary customizations have already happened by the time that it snapshots the environment as part of that first activation
- command: I think it would be very unusual for commands to be invoked in a child process of an existing activation, so much so that it we might even want to throw an error, but regardless again I think that the same "in-place" logic would apply here

It would be great to go through a battery of examples to see how this works in practice, and if we come up with any good ones not already represented in the test suite then we can add them there.  👍

Reply:
blocking (to at least discuss a little more): this also has some weird edge cases - right now, with `VAR=1` in `hook.on-activate`, `flox activate && export VAR=2 && eval "$(flox activate)"` results in `VAR=1`, but quite surprisingly if you `export VAR=1` first `export VAR=1 && flox activate && export VAR=2 && eval "$(flox activate)"` you end up with `VAR=2`. I'm not opposed to merging some weird edge cases that we can revisit, but I'm a little uneasy that the approach will ever be completely bug free. I'd want to either know we can fix it with a followup, or if we want to just accept the edge cases as the least bad option, I think maybe we should lay out the pros and cons a little more thoroughly first? I might think about that more tomorrow but wanted to wrap up my review before EOD

Reply:
I wonder if we could clear the environment or something before running the hook. I guess that wouldn't work well with variables that get appended like PATH

Reply:
Agree that there's a lot of potential for corner cases, and we need more real-world feedback to understand the sorts of things that are likely to occur in the wild.

> I wonder if we could clear the environment or something before running the hook

Interesting idea ... in discussion Matthew we discussed:
* the notion that the `activate` and the `hook.on-activate` script can / should be "pure", drawing only from packages in the environment
* thereafter the "profile" scripts can use whatever the userShell may introduce
* with this there's the option to a) capture the "background" environment, b) clear the environment prior to running `activate`, and c) overlay'ing the effects of the post-`activate` environment before continuing on with the userShell 

Bottom line this is destined to be an area we'll tweak into the future, so for today we should document the ordering of invocation and the fact that `activate` hooks run just once, and that "profile" scripts run with each new subshell, and thereafter just "lightly document" the specifics of how variables can be set, knowing this will be subject to change later.

Reply:
Got the verbal checkmark from Matthew on the above, resolving convo